gaia-header.js

```diff
@@ -145,13 +327,172 @@ module.exports = Component.register('gaia-header', {
    * @param  {Event} e
    * @private
    */
-  onActionButtonClick: function(e) {
-    var config = { detail: { type: this.getAttribute('action') } };
-    var actionEvent = new CustomEvent('action', config);
-    setTimeout(this.dispatchEvent.bind(this, actionEvent));
+  onActionButtonClick: function() {
+    debug('action button click');
+    var config = { detail: { type: this.action } };
+    var e = new CustomEvent('action', config);
+    setTimeout(() => this.dispatchEvent(e));
+  },
+
+  /**
+   * Get the margin required to center
+   * the title between surrounding buttons.
+   *
+   * We don't do any DOM writes in this function
+   * so that we can batch them together with
+   * other write for optimum performance.
+   *
+   * @param  {Object} title
+   * @return {Object}
+   * @private
+   */
+  getTitleMarginStart: function() {
+    var titleStart = this.titleStart;
+    var titleEnd = this.titleEnd;
+    debug('get title margin', titleStart, titleEnd);
+    return titleStart === titleEnd ? 0 : titleEnd - titleStart;
+  },
+
+  /**
+   * Get all the buttons (<a> & <button>)
+   * before the first <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsBeforeTitle: function() {
+    var els = [];
+
+    for (var i = 0; i < this.children.length; i++) {
+      var el = this.children[i];
+      if (el.tagName === 'H1') { break; }
+      if (!isButton(el.tagName)) { continue; }
+      els.push(el);
+    }
+
+    // Don't forget the action button
+    if (this.action) { els.push(this.els.actionButton); }
+    return els;
+  },
+
+  /**
+   * Get all the buttons (<a> & <button>)
+   * after the last <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsAfterTitle: function() {
+    var els = [];
+
+    for (var i = this.children.length - 1; i >= 0; i--) {
+      var el = this.children[i];
+      if (el.tagName === 'H1') { break; }
+      if (!isButton(el.tagName)) { continue; }
+      els.push(el);
+    }
+
+    return els;
+  },
```

Comment:
The user is able to omit the `action` attr and put one or more buttons before the title, so that is why we have to check for this.

The function below will ensure than hidden buttons do not contribute to overall `title-start` value.